server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1296,27 +1312,45 @@ public static String findV2Template(Metadata metadata, String indexName, boolean
      * one is the winner template that is applied to this index. In the event that no templates are matched,
-     * an empty list is returned.
+     * an empty list is returned. If <code>exitOnFirstMatch</code> is true, we return immediately after finding a match.
      */
-    static List<Tuple<String, ComposableIndexTemplate>> findV2CandidateTemplates(Metadata metadata, String indexName, boolean isHidden) {
+    static List<Tuple<String, ComposableIndexTemplate>> findV2CandidateTemplates(
+        Collection<Map.Entry<String, ComposableIndexTemplate>> templateEntries,
+        String indexName,
+        boolean isHidden,
+        boolean exitOnFirstMatch
+    ) {
         final String resolvedIndexName = IndexNameExpressionResolver.DateMathExpressionResolver.resolveExpression(indexName);
-        final Predicate<String> patternMatchPredicate = pattern -> Regex.simpleMatch(pattern, resolvedIndexName);
         final List<Tuple<String, ComposableIndexTemplate>> candidates = new ArrayList<>();
-        for (Map.Entry<String, ComposableIndexTemplate> entry : metadata.templatesV2().entrySet()) {
+        outerLoop:
+        for (Map.Entry<String, ComposableIndexTemplate> entry : templateEntries) {
             final String name = entry.getKey();
             final ComposableIndexTemplate template = entry.getValue();
             if (isHidden == false) {
-                final boolean matched = template.indexPatterns().stream().anyMatch(patternMatchPredicate);
-                if (matched) {
-                    candidates.add(Tuple.tuple(name, template));
+                for (String indexPattern : template.indexPatterns()) {
+                    if (Regex.simpleMatch(indexPattern, resolvedIndexName)) {
+                        candidates.add(Tuple.tuple(name, template));
+                        if (exitOnFirstMatch) {
+                            return candidates;
+                        }
+                    }
                 }
             } else {
-                final boolean isNotMatchAllTemplate = template.indexPatterns().stream().noneMatch(Regex::isMatchAllPattern);
-                if (isNotMatchAllTemplate) {
-                    if (template.indexPatterns().stream().anyMatch(patternMatchPredicate)) {
+                // For hidden indices, we don't want to return any templates that match "all".
+                boolean match = false;
+                for (String indexPattern : template.indexPatterns()) {
+                    if (Regex.isMatchAllPattern(indexPattern)) {
+                        continue outerLoop;
+                    }
+                    // If we've already found a match, we still want to keep looping to make sure this template doesn't have a "*" pattern.
+                    if (match == false && Regex.simpleMatch(indexPattern, resolvedIndexName)) {
                         candidates.add(Tuple.tuple(name, template));
+                        match = true;
                     }
                 }
```

Comment:
Is this performance boost of doing everything in one loop, worth it? This code is much harder to follow than the original that checked `isNotMatchAllTemplate` and it skipped.

Reply:
Similar as comment above, see [293ff9d](https://github.com/elastic/elasticsearch/pull/106953/commits/293ff9d7ecd187523221204c6f3c50920d51d011).